tests/src/test/scala/whisk/core/containerpool/docker/test/ApacheBlockingContainerClientTests.scala

```diff
@@ -106,20 +113,41 @@ class ContainerConnectionTests
 
   it should "handle empty entity response" in {
     val timeout = 5.seconds
-    val connection = new HttpUtils(hostWithPort, timeout, 1.B)
+    val connection = new ApacheBlockingContainerClient(hostWithPort, timeout, 1.B)
     testStatusCode = 204
-    val result = connection.post("/init", JsObject.empty, retry = true)
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
     result shouldBe Left(NoResponseReceived())
   }
 
+  it should "retry till timeout on HttpHostConnectException" in {
+    val timeout = 5.seconds
+    val badHostAndPort = "0.0.0.0:12345"
+    val connection = new ApacheBlockingContainerClient(badHostAndPort, timeout, 1.B)
+    testStatusCode = 204
+    val start = Instant.now()
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+    val end = Instant.now()
+    val waited = end.toEpochMilli - start.toEpochMilli
+    result should be('left)
+    result.left.get shouldBe a[Timeout]
+    result.left.get.asInstanceOf[Timeout].t shouldBe a[RetryableConnectionError]
+    result.left.get
+      .asInstanceOf[Timeout]
+      .t
+      .asInstanceOf[RetryableConnectionError]
+      .t shouldBe a[HttpHostConnectException]
```

Comment:
This could be rewritten a little less verbose like:

```scala
result match {
  case Left(Timeout(RetryableConnectionError(_: HttpHostConnectException))) => // all good
  case _ =>
    fail(s"$result was not a Timeout(RetryableConnectionError(HttpHostConnectException)))")
}
```

I was not able to reproduce this locally though, the test always failed with a ConnectError in both implementations. Is this intermittent?

Reply:
nice - much better! (worked for me)